src/libsystemd-network/sd-dhcp-server-lease.c

```diff
@@ -93,7 +98,10 @@ int dhcp_server_set_lease(sd_dhcp_server *server, be32_t address, DHCPRequest *r
 
         memcpy(lease->chaddr, req->message->chaddr, req->message->hlen);
 
-        if (req->hostname) {
+        /* Currently, the hostname field in the request is only used when dumping leases in JSON format, and
+         * the field is optional. So, let's silently ignore non UTF8 hostname. This needs to be updated when
+         * we really use the hostname field. */
+        if (req->hostname && utf8_is_valid(req->hostname)) {
```

Comment:
Interestingly the "Structure needs cleaning" thing was apparently introduced in https://github.com/systemd/systemd/pull/27817 and is reproducible with the main branch. I'm not sure why UTF-8 should be expected here. I wonder if it would make sense to dump bytes or something like that instead of using `JSON_BUILD_PAIR_STRING_NON_EMPTY` (where UTF-8 is enforced)

(It can probably be done later though)

Reply:
I opened https://github.com/systemd/systemd/issues/31708 so as not to forget about this.

Reply:
Thanks. Maybe, we should also add `HostnameRaw` field.

Reply:
The part loading the DHCP client leases apparently enforces UTF-8 too when it loads its leases (but it doesn't do it when it saves them):
```sh
# networkctl status host0 --no-pager                                                                                                                 
/run/systemd/netif/leases/2:13: invalid UTF-8 value for key ROOT_PATH: '��', ignoring.
...
```
I'm not exactly sure what to do about it but I can see that `fuzz-dhcp-server` now tries to save and then load its leases so the server part should be tested with all sorts of weird leases at least.

I'm not sure why CIFuzz is green but `fuzz-dhcp-server` fails to load the following lease (I'm guessing the hostname is the culprit there too):
```sh
{
  "BootID": "8fb3b8fae02a47adbe4052689d36652b",
  "Leases": [
    {
      "ClientId": [
        1,
        0,
        0,
        127,
        93,
        0,
        0
      ],
      "Address": [
        10,
        0,
        0,
        58
      ],
      "Hostname": "g\u0002=\u0002\u007f\u0002\u0002f\u0002\u0002e",
      "ExpirationUSec": 278876468747,
      "ExpirationRealtimeUSec": 1710103386715862
    },
    {
      "ClientId": [
        2,
        3
      ],
      "Address": [
        10,
        0,
        0,
        3
      ],
      "ExpirationUSec": 361676468686,
      "ExpirationRealtimeUSec": 1710186186715801
    },
    {
      "ClientId": [
        2,
        2
      ],
      "Address": [
        10,
        0,
        0,
        2
      ],
      "ExpirationUSec": 361676468605,
      "ExpirationRealtimeUSec": 1710186186715720
    }
  ]
}
```

Reply:
@evverx I have no idea about that. Maybe, there were garbage? Do you reproduce the issue reliably?

Reply:
The fuzz target crashes reliably. I sent a DHCPDISCOVER message with that hostname
```sh
    Option: (57) Maximum DHCP Message Size
        Length: 2
        Maximum DHCP Message Size: 1472
    Option: (80) Rapid commit
        Length: 0
        Value: <MISSING>
    Option: (12) Host Name
        Length: 11
        Host Name: g\002=\002\177\002\002f\002\002e
    Option: (255) End
        Option End: 255
```
and it broke `networkctl` at least:
```sh
networkctl status 30 --json=pretty
Failed to parse JSON: Invalid argument

networkctl status 30 |& grep -i lease
         Offered DHCP leases: 192.168.15.82 (to IAID:0x5888c0c1/DUID)
```
When networkd was restarted the lease got lost:
```
networkctl status 30 |& grep -i lease
         Offered DHCP leases: none
```
```
DHCPv4 server: Failed to load lease file /var/lib/systemd/network/dhcp-server-lease/ve-TEST-01, ignoring: Success
```

Reply:
For the record it's valid UTF-8 so it's different from https://github.com/systemd/systemd/issues/31708

Reply:
Found the root cause. Now the issue should be fixed.

Reply:
Looks like it's gone.

(Apparently it's reproducible with the main branch too so it wasn't introduced in this PR either).